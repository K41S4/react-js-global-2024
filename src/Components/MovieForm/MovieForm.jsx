import { useState } from 'react';
import styles from './MovieForm.module.css';
import { LabeledInput } from './LabeledInput/LabeledInput';

export const MovieForm = ({ initialValues = {}, onSubmit, genres }) => {
  const [movie, setMovie] = useState({
    title: initialValues?.title ?? '',
    releaseDate: initialValues?.releaseDate ?? '',
    imageUrl: initialValues?.imageUrl ?? '',
    rating: initialValues?.rating ?? '',
    genre: initialValues?.genre ?? '',
    runtime: initialValues?.runtime ?? '',
    description: initialValues?.description ?? '',
  });

  const handleChange = (e) => {
    const { name, value } = e.target;
    setMovie({ ...movie, [name]: value });
  };

  const handleReset = () => {
    setMovie({
      title: '',
      releaseDate: '',
      imageUrl: '',
      rating: '',
      genre: '',
      runtime: '',
      description: '',
    });
  };

  return (
    <form
      className={styles.container}
      onSubmit={(e) => {
        e.preventDefault();
        const formData = Object.fromEntries(new FormData(e.target));
        onSubmit(formData);
      }}
    >
      <div className={styles.formRow}>
        <div className={styles.formGroup}>
          <LabeledInput label={'Title'} onChange={handleChange} value={movie.title} id={'title'} />
        </div>
        <div className={styles.formGroup}>
          <LabeledInput label={'Release Date'} onChange={handleChange} value={movie.releaseDate} id={'releaseDate'} />
        </div>
      </div>
      <div className={styles.formRow}>
        <div className={styles.formGroup}>
          <LabeledInput label={'Image Url'} onChange={handleChange} value={movie.imageUrl} id={'imageUrl'} />
        </div>
        <div className={styles.formGroup}>
          <LabeledInput label={'Rating'} onChange={handleChange} value={movie.rating} id={'rating'} />
        </div>
      </div>
      <div className={styles.formRow}>
        <div className={styles.formGroup}>
          <label className={styles.label} htmlFor={'genres'}>
            Genre
          </label>
          <select className={styles.select} id={'genres'} name="genre" defaultValue={movie.genre}>
            <option hidden></option>
            {!!genres?.length &&
              genres.map((genre) => (
                <option key={genre} value={genre}>
                  {genre}
                </option>
              ))}
          </select>
        </div>
        <div className={styles.formGroup}>
          <LabeledInput label={'Runtime'} onChange={handleChange} value={movie.runtime} id={'runtime'} />
        </div>
      </div>

      <div className={styles.formGroup}>
        <label htmlFor="description" className={styles.overviewLabel}>
          Description
        </label>
        <textarea
          id="description"
          name="description"
          className={styles.overviewTextarea}
          rows="4"
          value={movie.description}
          onChange={handleChange}
        ></textarea>
      </div>

      <div className={styles.formActions}>
        <button className={styles.resetButton} type="reset" onClick={handleReset}>
          Reset
        </button>
        <button className={styles.submitButton}>Submit</button>
      </div>
    </form>
  );
};
