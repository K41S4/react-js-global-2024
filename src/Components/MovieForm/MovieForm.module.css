.formRow {
  display: flex;
  justify-content: space-between;
  gap: 20px;
}

.formGroup {
  display: flex;
  flex-direction: column;
  flex: 1;
}

.label {
  margin-bottom: 5px;
  color: #f65261;
}

.select {
  padding: 10px;
  margin-bottom: 20px;
  background-color: #424242;
  border: none;
  border-radius: 2px;
  color: #ffffff;
}

.overviewTextarea {
  max-width: calc(100% - 20px);
  background-color: #424242;
  color: #ffffff;
  border: none;
  padding: 10px;
  margin-bottom: 20px;
}

.formActions {
  display: flex;
  justify-content: end;
}

.resetButton {
  background-color: #232323;
  color: #f65261;
  border-color: #f65261;
  border-width: 1px;
  border-style: solid;
  padding: 10px 40px;
  margin-right: 20px;
}

.submitButton {
  background-color: #f65261;
  color: #ffffff;
  border: none;
  padding: 10px 40px;
}
