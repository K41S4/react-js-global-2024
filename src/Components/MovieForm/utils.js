export const validateForm = (values) => {
  const errors = {};
  if (!values.title.length) {
    errors.title = 'Title is required';
  }
  if (!values.description.length) {
    errors.description = 'Description is required';
  }
  if (!values.genre.length) {
    errors.genre = 'Genre is required';
  }
  if (!values.releaseYear.length) {
    errors.releaseYear = 'Release date is required';
  } else if (!isValidDate(values.releaseYear)) {
    errors.releaseYear = 'Release date must be valid';
  }

  if (!values.imageUrl.length) {
    errors.imageUrl = 'Image URL is required';
  } else if (!isValidUrl(values.imageUrl)) {
    errors.imageUrl = 'Image URL must be valid URL';
  }

  if (values.rating > 10 || values.rating < 0) {
    errors.rating = 'Rating must be within 0-10 range';
  }

  return errors;
};

const isValidDate = (input) => {
  const date = new Date(input);
  return !isNaN(date.getTime()) && input === date.toISOString().slice(0, 10);
};
const isValidUrl = (url) => {
  try {
    new URL(url);
    return true;
  } catch (e) {
    return false;
  }
};
