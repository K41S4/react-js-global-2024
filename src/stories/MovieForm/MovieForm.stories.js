import { fn } from '@storybook/test';

import { MovieForm } from '../../Components/MovieForm/MovieForm';

export default {
  title: 'Example/MovieForm',
  component: MovieForm,
  parameters: {
    layout: 'centered',
    backgrounds: {
      default: 'dark',
    },
  },
  tags: ['autodocs'],
  args: {
    onSubmit: fn(),
  },
};

export const MovieFormDefault = {
  args: {
    genres: ['All', 'Horror', 'Crime'],
    initialValues: {
      title: '',
      releaseDate: '',
      imageUrl: '',
      rating: '',
      genre: '',
      runtime: '',
      description: '',
    },
  },
};
