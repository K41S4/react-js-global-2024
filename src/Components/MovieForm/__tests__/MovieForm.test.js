import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MovieForm } from '../MovieForm';

describe('MovieForm', () => {
  it('allows entering values into form fields', async () => {
    const user = userEvent.setup();
    const movieDetails = {
      title: 'Test title',
      releaseDate: '1999',
      imageUrl: 'http://image',
      rating: '9',
      genre: 'Action',
      runtime: '1h',
      description: 'Test description',
    };

    render(<MovieForm onSubmit={jest.fn()} genres={['Action']} />);

    const titleInput = screen.getByLabelText('Title');
    const releaseDateInput = screen.getByLabelText('Release Date');
    const imageUrlInput = screen.getByLabelText('Image Url');
    const ratingInput = screen.getByLabelText('Rating');
    const genreInput = screen.getByLabelText('Genre');
    const runtimeInput = screen.getByLabelText('Runtime');
    const descriptionTextarea = screen.getByLabelText('Description');

    await user.type(titleInput, movieDetails.title);
    await user.type(releaseDateInput, movieDetails.releaseDate);
    await user.type(imageUrlInput, movieDetails.imageUrl);
    await user.type(ratingInput, movieDetails.rating);
    await user.selectOptions(genreInput, movieDetails.genre);
    await user.type(runtimeInput, movieDetails.runtime);
    await user.type(descriptionTextarea, movieDetails.description);

    expect(titleInput).toHaveValue(movieDetails.title);
    expect(releaseDateInput).toHaveValue(movieDetails.releaseDate);
    expect(imageUrlInput).toHaveValue(movieDetails.imageUrl);
    expect(ratingInput).toHaveValue(movieDetails.rating);
    expect(genreInput).toHaveValue(movieDetails.genre);
    expect(runtimeInput).toHaveValue(movieDetails.runtime);
    expect(descriptionTextarea).toHaveValue(movieDetails.description);
  });

  it('handles submit event with correct values', async () => {
    const user = userEvent.setup();
    const mockSubmit = jest.fn();
    const title = 'test title';
    render(<MovieForm onSubmit={mockSubmit} genres={['']} />);

    await user.type(screen.getByLabelText('Title'), title);
    await user.click(screen.getByRole('button', { name: 'Submit' }));

    expect(mockSubmit).toHaveBeenCalledWith({
      title: title,
      releaseDate: '',
      imageUrl: '',
      rating: '',
      genre: '',
      runtime: '',
      description: '',
    });
  });

  it('resets the form when reset button is clicked', async () => {
    const user = userEvent.setup();

    render(<MovieForm onSubmit={jest.fn()} genres={[]} />);

    const titleInput = screen.getByLabelText('Title');

    await user.type(titleInput, 'test');
    expect(titleInput).toHaveValue('test');

    await user.click(screen.getByRole('button', { name: 'Reset' }));
    expect(titleInput).toHaveValue('');
  });

  it('sets initial form values', async () => {
    const genres = ['Action', 'Drama', 'Comedy'];

    const movieDetails = {
      title: 'Test title',
      releaseDate: '1999',
      imageUrl: 'http://image',
      rating: '9',
      genre: 'Action',
      runtime: '1h',
      description: 'Test description',
    };

    const { asFragment } = render(<MovieForm initialValues={movieDetails} onSubmit={jest.fn()} genres={genres} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
