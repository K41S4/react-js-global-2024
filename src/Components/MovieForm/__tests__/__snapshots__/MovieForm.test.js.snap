// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieForm sets initial form values 1`] = `
<DocumentFragment>
  <form
    class="container"
  >
    <div
      class="formRow"
    >
      <div
        class="formGroup"
      >
        <label
          class="label"
          for="title"
        >
          Title
        </label>
        <input
          class="input"
          id="title"
          name="title"
          type="text"
          value="Test title"
        />
      </div>
      <div
        class="formGroup"
      >
        <label
          class="label"
          for="releaseDate"
        >
          Release Date
        </label>
        <input
          class="input"
          id="releaseDate"
          name="releaseDate"
          type="text"
          value="1999"
        />
      </div>
    </div>
    <div
      class="formRow"
    >
      <div
        class="formGroup"
      >
        <label
          class="label"
          for="imageUrl"
        >
          Image Url
        </label>
        <input
          class="input"
          id="imageUrl"
          name="imageUrl"
          type="text"
          value="http://image"
        />
      </div>
      <div
        class="formGroup"
      >
        <label
          class="label"
          for="rating"
        >
          Rating
        </label>
        <input
          class="input"
          id="rating"
          name="rating"
          type="text"
          value="9"
        />
      </div>
    </div>
    <div
      class="formRow"
    >
      <div
        class="formGroup"
      >
        <label
          class="label"
          for="genres"
        >
          Genre
        </label>
        <select
          class="select"
          id="genres"
          name="genre"
        >
          <option
            hidden=""
          />
          <option
            selected=""
            value="Action"
          >
            Action
          </option>
          <option
            value="Drama"
          >
            Drama
          </option>
          <option
            value="Comedy"
          >
            Comedy
          </option>
        </select>
      </div>
      <div
        class="formGroup"
      >
        <label
          class="label"
          for="runtime"
        >
          Runtime
        </label>
        <input
          class="input"
          id="runtime"
          name="runtime"
          type="text"
          value="1h"
        />
      </div>
    </div>
    <div
      class="formGroup"
    >
      <label
        class="overviewLabel"
        for="description"
      >
        Description
      </label>
      <textarea
        class="overviewTextarea"
        id="description"
        name="description"
        rows="4"
      >
        Test description
      </textarea>
    </div>
    <div
      class="formActions"
    >
      <button
        class="resetButton"
        type="reset"
      >
        Reset
      </button>
      <button
        class="submitButton"
      >
        Submit
      </button>
    </div>
  </form>
</DocumentFragment>
`;
